src/Core/Injector/Injector.php

```diff
@@ -581,6 +582,11 @@ protected function instantiate($spec, $id = null, $type = null)
             $constructorParams = $spec['constructor'];
         }
 
+        // If we're dealing with a DataObject, pass through Singleton flag as second argument
+        if ($type != 'prototype' && empty($constructorParams) && is_subclass_of($class, DataObject::class)) {
```

Comment:
Strict comparison, please. Also use the provided constant (`self::PROTOTYPE`)

Reply:
Fixed. Also checked for !$type, as that's interpreted as singleton elsewhere.